crates/tauri-schema-worker/src/config.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2019-2024 Tauri Programme within The Commons Conservancy
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
+
+use anyhow::Context;
+use axum::{
+  extract::Path,
+  http::{header, StatusCode},
+  response::Result,
+  routing::get,
+  Router,
+};
+use semver::{Version, VersionReq};
+use serde::Deserialize;
+use worker::*;
+
+#[derive(Deserialize)]
+pub struct CrateReleases {
+  pub versions: Vec<CrateRelease>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct CrateRelease {
+  #[serde(alias = "num")]
+  pub version: Version,
+  pub yanked: Option<bool>,
+}
+
+#[derive(Deserialize)]
+pub struct CrateMetadataFull {
+  #[serde(rename = "crate")]
+  pub crate_: CrateMetadata,
+}
+
+#[derive(Deserialize)]
+pub struct CrateMetadata {
+  pub max_stable_version: Version,
+}
+
+const USERAGENT: &str = "tauri-schema-worker (contact@tauri.app)";
+
+pub fn router() -> Router {
+  Router::new()
+    .route("/config", get(latest_stable_schema))
+    .route("/config/latest", get(latest_stable_schema))
+    .route("/config/stable", get(latest_stable_schema))
+    .route("/config/next", get(latest_next_schema)) // pre-releases versions, (rc, alpha and beta)
+    .route("/config/:version", get(schema))
+}
+
+async fn schema(Path(version): Path<String>) -> Result<String> {
+  Ok(
+    try_schema(version)
+      .await
+      .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?,
+  )
+}
+
+async fn latest_stable_schema() -> Result<String> {
+  Ok(
+    max_stable_schema()
+      .await
+      .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?,
+  )
+}
+
+async fn latest_next_schema() -> Result<String> {
+  Ok(
+    try_latest_prerelease_schema()
+      .await
+      .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?,
+  )
+}
+
+#[worker::send]
+async fn try_latest_prerelease_schema() -> anyhow::Result<String> {
+  let releases = crate_releases("tauri").await?;
+  let version = releases
+    .into_iter()
+    .filter(|r| !r.version.pre.is_empty())
+    .map(|r| r.version)
+    .max()
+    .context("Couldn't find latest pre-release")?;
+  schema_file_for_version(version).await
+}
+
+#[worker::send]
+async fn try_schema(version: String) -> anyhow::Result<String> {
+  let version = version.parse::<VersionReq>()?;
+
+  let releases = crate_releases("tauri").await?;
+
+  if releases.is_empty() {
+    return max_stable_schema().await;
+  }
+
+  let Some(version) = releases.into_iter().find(|r| version.matches(&r.version)) else {
+    return max_stable_schema().await;
+  };
+
+  schema_file_for_version(version.version).await
+}
+
+async fn crate_releases(crate_: &str) -> anyhow::Result<Vec<CrateRelease>> {
```

Comment:
what do you think about caching these for a while? maybe in the future even add a webhook for tauri publishes?

Reply:
that's a good idea, will have to read more about workers 😅

Reply:
maybe the worker can cache the JSON schema per version, then when we request `/config/2` we just redirect to the latest version `/config/2.1.2` or whatever, so we don't really need to deal with it
David knows how to do it, he wrote some cloudflare caching in the past

Reply:
`/config/2` is resolved to `2.1.2` or whatever latest is, and response will be cached for resolved version so `/config/2` or `/config/2.1` or `/config/2.1.2` or `/config/latest` (and other aliases) will use the same file from cache.